trl/trainer/kto_trainer.py

```diff
@@ -326,78 +480,157 @@ def make_inputs_require_grad(module, input, output):
+            )
 
-        # split the dataset and interleave them together with equal probability of choosing chosen or rejected
-        interleaved_train_dataset = interleave_datasets(
-            [desirable, undesirable],
-            stopping_strategy="all_exhausted",
-        )
-        interleaved_train_dataset = interleaved_train_dataset.shuffle(seed=args.data_seed)
+            if len(desirable) != len(undesirable):
+                # The lower and upper bounds come from Eq. (8) of https://arxiv.org/abs/2402.01306
+                des_weight_lower_bound = round((len(undesirable) * self.undesirable_weight / len(desirable)) * 1, 2)
+                des_weight_upper_bound = round((len(undesirable) * self.undesirable_weight / len(desirable)) * 1.33, 2)
+                und_weight_lower_bound = round((len(desirable) * self.desirable_weight / len(undesirable)) / 1.33, 2)
+                und_weight_upper_bound = round((len(desirable) * self.desirable_weight / len(undesirable)) / 1, 2)
+
+                des_weight_in_range = des_weight_lower_bound <= self.desirable_weight <= des_weight_upper_bound
+                und_weight_in_range = und_weight_lower_bound <= self.undesirable_weight <= und_weight_upper_bound
+
+                if not (des_weight_in_range or und_weight_in_range):
+                    warnings.warn(
+                        f"""
+                        You have different amounts of desirable/positive and undesirable/negative examples but the
+                        weights on the desirable and undesirable losses don't seem to be in an ideal range. Based
+                        on your data, we recommend EITHER desirable_weight in [{des_weight_lower_bound}, {des_weight_upper_bound}]
+                        or undesirable_weight in [{und_weight_lower_bound}, {und_weight_upper_bound}] (but NOT BOTH).
+                        See the documentation on how to optimally set these weights.""",
+                        UserWarning,
+                    )
 
-        if eval_dataset is not None:
-            interleaved_eval_dataset = interleave_datasets(
-                [eval_dataset.filter(lambda x: x["label"]), eval_dataset.filter(lambda x: not x["label"])],
+            # split the dataset and interleave them together with equal probability of choosing chosen or rejected
+            interleaved_train_dataset = interleave_datasets(
+                [desirable, undesirable],
                 stopping_strategy="all_exhausted",
             )
-        else:
-            interleaved_eval_dataset = None
+            interleaved_train_dataset = interleaved_train_dataset.shuffle(seed=args.data_seed)
+
+            if eval_dataset is not None:
+                interleaved_eval_dataset = interleave_datasets(
+                    [
+                        eval_dataset.filter(lambda x: x["label"], num_proc=args.dataset_num_proc),
```

Comment:
I am aware of #1476 which removes the interleaving, so will rebase after that PR is merged to make sure we don't re-introduce this step